### PR TITLE
DROOLS-1531 Fix instanceof accordingly to FEEL spec table 49 (optimizations)

### PR DESCRIPTION
**Pull request by tarilabs**
@etirelli please consider this PR for the 2 remaining comments which were noted here: https://github.com/kiegroup/drools/pull/1214#pullrequestreview-34157657

### CODE REVIEW

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java [hunk 1]
```diff
@@ -47,19 +47,14 @@ public MapBackedType addField(String name, Type type) {
...
-            instance = (Map<String, Object>) o;
-        } catch ( Exception e ) {
+        if ( o == null || !(o instanceof Map) ) {
             return false;
         }
+        Map<?, ?> instance = (Map<?, ?>) o;
         for ( Entry<String, Type> f : fields.entrySet() ) {
-            if ( instance.get(f.getKey()) == null ) {
+            Object instanceValueForKey = instance.get(f.getKey());
+            if ( instance.get(f.getKey()) == null || !f.getValue().isInstanceOf(instanceValueForKey) ) {
```
Comment (etirelli) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java]: @tarilabs is this correct? There are two cases, when a map backed instance does not have an a given attribute, and when it has the attribute but the value is null.
For instance, a Person instance might require the "Last Name" attribute to be considered a person, but the attribute might be set to null. It is still a person if the last name is null, but in duck typing it is not a person if the last name attribute does not exist.

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java [hunk 2]
```diff
@@ -70,19 +65,14 @@ public boolean isAssignableValue(Object value) {
...
-            instance = (Map<String, Object>) value;
-        } catch ( Exception e ) {
+        if ( !(value instanceof Map) ) {
             return false;
         }
+        Map<?, ?> instance = (Map<?, ?>) value;
         for ( Entry<String, Type> f : fields.entrySet() ) {
-            if ( instance.get(f.getKey()) == null ) {
+            Object instanceValueForKey = instance.get(f.getKey());
+            if ( instanceValueForKey == null || !f.getValue().isAssignableValue(instanceValueForKey) ) {
```
Comment (etirelli) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java]: same here

**Comment by tarilabs** [comment]
@etirelli you are absolutely right, I'm just a little lost as I thought this is what I did originally with the two separate if checks, possibly I lost it even before during some refactoring.. checking.

**Comment by tarilabs** [comment]
@etirelli I've updated following your comments, and the new test manifested an issue which was present on master even before (https://github.com/kiegroup/drools/blob/master/kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java#L30) , now corrected.

**Review by etirelli — APPROVED** [review]

**Comment by etirelli** [comment]
Looks good @tarilabs ! Thanks!

**Pull request merged**